### PR TITLE
Add support for GL_EXT_shared_memory_block

### PR DESCRIPTION
**Pull request by cmarcelo**
This will use generate SPIR-V that uses SPV_KHR_workgroup_memory_explicit_layout extension.

The first patch is a fix for #2400 that I can move to a PR of its own if necessary. It is here because we build on top of that in a later patch.

### CODE REVIEW

File: known_good.json [hunk 1]
```diff
@@ -5,14 +5,14 @@
       "site" : "github",
       "subrepo" : "KhronosGroup/SPIRV-Tools",
       "subdir" : "External/spirv-tools",
-      "commit" : "ad898cb9498c65cac7c5b6f9ebddb610bd0852ad"
+      "commit" : "28b7ca928692ecaaa8558357f9f9ce66e14ac8ef"
```
Comment (cmarcelo) [inline known_good.json:8]: NOTE: This will be updated once https://github.com/KhronosGroup/SPIRV-Tools/pull/4128 lands.

File: known_good.json [hunk 2]
```diff
@@ -5,14 +5,14 @@
...
     },
     {
       "name" : "spirv-tools/external/spirv-headers",
       "site" : "github",
       "subrepo" : "KhronosGroup/SPIRV-Headers",
       "subdir" : "External/spirv-tools/external/spirv-headers",
-      "commit" : "f027d53ded7e230e008d37c8b47ede7cd308e19d"
+      "commit" : "38fe9b08508c876ba830f250a488c51003531e7e"
```
Comment (cmarcelo) [inline known_good.json:15]: NOTE: This will be updated once https://github.com/KhronosGroup/SPIRV-Headers/pull/192 lands.

**Comment by cmarcelo** [comment]
@greg-lunarg could you take a look at this?

**Comment by greg-lunarg** [comment]
I am not seeing 28b7ca as a valid commit for SPIRV-Tools on github. This new value in known_good.json appears to be the reason for the CI failures. You indicate that you are going to update these anyway once SPIRV-Tools and SPIRV-Headers are updated. Once valid github commits are put in here, things should go better.

**Comment by cmarcelo** [comment]
> I am not seeing 28b7ca as a valid commit for SPIRV-Tools on github. This new value in known_good.json appears to be the reason for the CI failures. You indicate that you are going to update these anyway once SPIRV-Tools and SPIRV-Headers are updated. Once valid github commits are put in here, things should go better.

@greg-lunarg Correct, these are the hashes of my PRs, but they don't get propagated in a regular clone. SPIRV-Headers already merged and SPIRV-Tools is in the process, I'll update this PR once the latter lands.

I did push this PR earlier because some of the new changes can be reviewed already, e.g. the first patch that fixes #2400. I can also move it to a separate PR of its own if it helps, just let me know.

**Comment by greg-lunarg** [comment]
I looked at most of this already but would like to look at the #2400 fix.

**Comment by alan-baker** [comment]
I merged the spirv-tools change, but it also includes validation changes for workgroup variable initializers and stricter validation for offset. The test expectations in glslang will need updated as a result.

**Comment by cmarcelo** [comment]
Thanks Alan. I'll try to update the test expectations and include that in this PR.

File: known_good.json (hunk 1 above)
Comment (cmarcelo) [reply known_good.json]: Updated.

File: known_good.json (hunk 2 above)
Comment (cmarcelo) [reply known_good.json]: Updated.

**Comment by cmarcelo** [comment]
@alan-baker @greg-lunarg updated.

**Comment by greg-lunarg** [comment]
As a whole, these changes look good.

Do we need to squash some of these? I just want to make sure we don't leave points at which the repo is not coherent.

**Comment by cmarcelo** [comment]
> As a whole, these changes look good.
> 
> Do we need to squash some of these? I just want to make sure we don't leave points at which the repo is not coherent.

Updated. I've squashed:

- The known_good and the copy of header update, since kind of make sense.
- The three last patches.

**Pull request merged**